src/arch/linux/async/signal.c

```diff
@@ -1165,6 +1165,11 @@ static void sigasync0(int sig, sigcontext_t *scp, siginfo_t *si)
 {
   pthread_t tid = pthread_self();
   if (!pthread_equal(tid, dosemu_pthread_self)) {
+    /* Nvidia libGL creates a thread ("CPMMListener") which sends
+     * SIGVTALARM to itself. Don't bother spawning gdb for a backtrace.
+     */
+    if (sig == SIGVTALRM || sig == SIGALRM)
```

Comment:
I am not sure this is the right thing
to do. AFAIK the signals are per-process,
not per-thread. So if the nvidia thread
uses that signal and we use it too,
it won't work.
And by just doing a "return" here
you only avoid the crash, but you
don't do anything for the nvidia
thread to actually get that signal.

So I believe, there are following
problems:
- nvidia handler is not going to be executed
- signal randomly come either to our
thread (then we (mis)treat it as ours)
or to nvidia thread (where you meet
it with plain return).

Using signals in some random thread,
is a very bad thing to do. nvidia just
shouldn't do that, or there should be
an env var to stop it from doing that.

Reply:
Hmm you are right, they are per process. Which is why we see them.

How about storing old signal handlers when we register ours and invoke them when its not a thread we care about?

What bothers me is that I can't find anyone else complaining about this online. If the driver does this all the time someone else should have seen it.

glxgears on Nvidia doesn't have this CPMMListener thread and doesn't see any SIGVTALRM and SIGALRM signals.

Reply:
> How about storing old signal handlers when we register ours and invoke them when its not a thread we care about?

Likely won't work, for many reasons.
The ones from the top of my head:
- nvidia lib gets dlopened together with
sdl, when all signals are already set up
- you will still not distinguish "our" vs
"nvidia" signals.
Probably there are more.

Reply:
I think it is our own SIGVTALRM that gets sent to Nvidia's thread: https://stackoverflow.com/questions/2586926/setitimer-sigalrm-multithread-process-linux-c

Sure enough, disabling the cpu-emu.c setitimer call eliminates the weird VTALRM in that Nvidia driver thread.

Why though does it never hit any other thread, like the two async render threads?

Reply:
> Sure enough, disabling the cpu-emu.c setitimer call eliminates the weird VTALRM in that Nvidia driver thread.

So its our bug then?
OK, looking...

Reply:
Well, our threads inherit the
sigprocmask, which is per-thread.
The problem here is that nvidia
thread is started too late - when
we already released mask from
the main thread, which would
otherwise be inherited from the
nvidia thread, and no problem
will happen.

I thought about the simple solution:
like for example set the mask when
dlopen()ing plugin, and release it
after.
But unfortunately this still doesn't
guarantee the nvidia code to create
the thread at the right point.
Still, I highly suspect the nvidia
code creates that thread at a dlopen()
time, rather than any later, because
"any later" will mean "from opengl
API calls".

So... would you like to try the solution
with temporarily setting the sigprocmask
upon dlopen?

Reply:
No, not dlopen(), but rather SDL_Init()
or SDL_InitSubsystem() or alike.
So you need to find which SDL call
creates the thread, and manipulate
the sigprocmask with pthread_sigmask()
around that call.

Reply:
(gdb) info threads
  Id   Target Id                  Frame 
* 1    LWP 346585 "dosemu.bin"    0x00007fc400c66bea in sigsuspend () from /lib64/libc.so.6
  2    LWP 346604 "dosemu: sdl_r" 0x00007fc400f680ea in __futex_abstimed_wait_common64 () from /lib64/libpthread.so.0
  3    LWP 346605 "PulseHotplug"  0x00007fc400d1c1d6 in ppoll () from /lib64/libc.so.6
  4    LWP 346606 "SDLAudioP2"    0x00007fc400d1c1d6 in ppoll () from /lib64/libc.so.6
  5    LWP 346612 "dosemu: adlib" 0x00007fc400f680ea in __futex_abstimed_wait_common64 () from /lib64/libpthread.so.0
  6    LWP 346613 "CPMMListener"  0x00007fc400d1e739 in select () from /lib64/libc.so.6


So I have at least 3 library-created threads, two created by SDL for Audio stuff and one created by Nvidia. I don't think wrapping whatever calls might create a thread is going to work. Lib behaviour might change in the future and users might inject third party stuff, e.g. things similar to OBS or the Steam overlay to record their l33t DOS gaming.

My guess is setitimer is one of those ancient POSIX APIs that didn't age well.

The setitimer man page suggests timer_create instead. It has SIGEV_THREAD_ID, which is Linux specific, but if you dabble in KVM, vm86 and modify_ldt, a Linux specific timer API is the least of your worries.

Alternatively pthread_kill, possibly combined with blocking the signal in the thread for the future? Not sure if that has its own set of issues.